core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -83,6 +87,8 @@
       ScheduledExecutorService deadlineCancellationExecutor) {
     this.method = method;
     this.callExecutor = executor;
+    // Propagate the context from the thread which initiated the call to all callbacks.
+    this.context = Context.current();
```

Comment:
I think that's a mistake in the docs ([left over](https://github.com/grpc/grpc-java/commit/e4bd1c7d69e62fa87250cc4bf4bd894bb3d51359) from very old semantic tweaks). I am very certain we will close without any onHeaders in error conditions, and I would believe we would want to continue. I would claim "onHeaders is always called before onPayload".